ocaml/sdk-gen/csharp/autogen/src/Marshalling.cs

```diff
@@ -108,14 +106,86 @@ public static long[] ParseLongArray(Hashtable table, string key)
 
         public static List<XenRef<T>> ParseSetRef<T>(Hashtable table, string key) where T : XenObject<T>
         {
-            var rs = (object[])table[key];
-            return rs == null ? null : XenRef<T>.Create(rs);
+            return ParseSxpList((string)table[key]).Select(XenRef<T>.Create).ToList();
+        }
+
+
+        private static Hashtable ParseSxpDict(string p)
+        {
+            var result = new Hashtable();
+
+            using (var enumerator = Tokenize(p).GetEnumerator())
+            {
+                if (!enumerator.MoveNext())
+                    return result;
+
+                while (enumerator.MoveNext())
+                {
+                    if (enumerator.Current == ")")
+                        break;
+
+                    enumerator.MoveNext();
+                    var key = enumerator.Current;
+                    enumerator.MoveNext();
+                    var value = enumerator.Current;
+                    enumerator.MoveNext();
+
+                    result[key] = value;
+                }
+
+                return result;
+            }
+        }
+
+        private static List<string> ParseSxpList(string p)
+        {
+            var result = new List<string>();
+
+            foreach (var token in Tokenize(p))
+            {
+                if (token == "(" || token == ")")
+                    continue;
+
+                result.Add(token);
+            }
+
+            return result;
         }
 
-        public static Dictionary<XenRef<T>, T> ParseMapRefRecord<T, U>(Hashtable table, string key) where T : XenObject<T>
+        private static IEnumerable<string> Tokenize(string str)
         {
-            Hashtable map = ParseHashTable(table, key);
-            return map == null ? null : XenRef<T>.Create<U>(map);
+            bool inStr = false;
+            int j = 0;
+
+            for (int i = 0; i < str.Length; i++)
+            {
+                switch (str[i])
+                {
+                    case '(':
+                        if (!inStr)
+                            yield return "(";
+                        break;
+                    case ')':
+                        if (!inStr)
+                            yield return ")";
+                        break;
+
+                    case '\'':
+                    case '"':
+                        if (!inStr)
+                        {
+                            inStr = true;
+                            j = i;
+                        }
+                        else if (str[i - 1] != '\\')
+                        {
+                            inStr = false;
+                            yield return str.Substring(j + 1, i - j - 1).Replace("\\\"", "\"").Replace("\\\'", "\'");
+                        }
+
+                        break;
+                }
+            }

```

Comment:
The high level method that calls this one is the API objects' UpdateFrom(Hashtable). This is used (1) when the client loads directly a xapi database, (2) when the PoSh SDK creates an object from a Hashtable, and (3) when the client imports an OVF and parses xenserver specific settings. The values for each API object's property are parsed by the methods called within the UpdateFrom. The tokenization serves for sanitizing the weird strings contained within the xapi db, in the other cases the input is "cleaner" and the tokenization should not affect it (at least it did not affect it with the basic dev testing I did, we may want to run more tests, especially for the PoSh case as construction from hashtables is more obscure that the others).

Reply:
what sort of testing? I can see this breaking with some weird corner cases

Reply:
In the loop above, what happens with `str[i - 1]` when `I == 0` and `str[0] == '\''`? Can you exclude that this can happen?

Reply:
If i==0 it is the first time the loop is executed, so `inStr` is false and it won't hit the `str[i-1]` line.

I did some further testing with the PS module and it turned out some alterations were needed. I added a new commit.

Reply:
@danilo-delbusso  the new commit should fix the non-string values. Regarding unhelful messages, I'm not very bothered because I don't expect customer uses to access this code very frequently.